### PR TITLE
sagemaker_domain add docker_settings & fix domain_settings update

### LINKED ISSUE #31912: [Bug]: aws_sagemaker_domain ignores API errors
### Terraform Core Version

v1.3.7

### AWS Provider Version

v5.2.0

### Affected Resource(s)

* aws_sagemaker_domain

### Expected Behavior

Terraform should either succeed applying settings or fail with error from upstream API

### Actual Behavior

Terraform apply completes without issues, but the settings are not actually applied. Next plan shows them again.

### Relevant Error/Panic Output Snippet

```shell
The awscli command shows the actual error.


$ aws sagemaker \
>     update-domain \
>     --domain-id d-xxxxxxxxxxxx \
>     --app-security-group-management Service \
>     --domain-settings-for-update "RStudioServerProDomainSettingsForUpdate={DomainExecutionRoleArn=arn:aws:iam::123456789:role/xxxxxxxxxxx,RStudioConnectUrl=https://posit-connect.example.com/connect/,RStudioPackageManagerUrl=https://posit-pm.example.com/cran/latest}"

An error occurred (ValidationException) when calling the UpdateDomain operation: Unable to update Domain [arn:aws:sagemaker:us-east-1:123456789:domain/d-xxxxxxxxxxxx] with InService App. Delete all InService apps in the domain including shared Apps for [domain-shared] User Profile, and try again.
```

### Terraform Configuration Files

```hcl
resource "aws_sagemaker_domain" "this" {
  domain_name             = local.name
  auth_mode               = "IAM"
  vpc_id                  = data.aws_vpc.xxx.id
  subnet_ids              = data.aws_subnets.xxx.ids
  app_network_access_type = "VpcOnly"

  app_security_group_management = "Service"

  domain_settings {
    security_group_ids = []

    r_studio_server_pro_domain_settings {
      domain_execution_role_arn    = module.rstudio-domain-execution-role.iam_role_arn
      r_studio_connect_url         = "https://posit-connect.example.com/connect/"
      r_studio_package_manager_url = "https://posit-pm.example.com/cran/latest"
    }
  }
}
```

### Steps to Reproduce

1. Create domain without rstudio settings
2. Update config with rstudio settings
3. Terraform applies without error, but the changes are actually not applied.

### Debug Output

_No response_

### Panic Output

_No response_

### Important Factoids

_No response_

### References

_No response_

### Would you like to implement a fix?

No

### LINKED ISSUE #35115: [Enhancement]: SageMaker Studio now supports local mode and Docker
### Description

Studio users can now run SageMaker processing, training, inference and batch transform jobs locally on their Studio IDE instance. Users can also build and test SageMaker compatible Docker images locally in Studio IDEs. [link](https://aws.amazon.com/about-aws/whats-new/2023/12/sagemaker-studio-local-mode-docker/)

Usage guide - [Link](https://docs.aws.amazon.com/sagemaker/latest/dg/studio-updated-local.html)

@DrFaust92 

### Affected Resource(s) and/or Data Source(s)

aws_sagemaker_domain

### Potential Terraform Configuration

```terraform
resource "aws_sagemaker_domain" "example" {
  domain_name = "example"
  auth_mode   = "IAM"
  vpc_id      = aws_vpc.example.id
  subnet_ids  = [aws_subnet.example.id]

  default_user_settings {
    execution_role  = aws_iam_role.example.arn
    studio_web_portal = "ENABLED"
    kernel_gateway_app_settings {
      custom_image {
        app_image_config_name = aws_sagemaker_app_image_config.example.app_image_config_name
        image_name            = aws_sagemaker_image_version.example.image_name
      }
    }
  }

    domain_settings {
      execution_role_identity_config = "USER_PROFILE_NAME|DISABLED"
      docker_settings {
        enable_docker_access      = "ENABLED|DISABLED"
        vpc_only_trusted_accounts = ["123456789012","111111111111"]
      }
    }  
}
```

### References

Code Reference: https://github.com/hashicorp/terraform-provider-aws/blob/main/internal/service/sagemaker/domain.go#L868

API Reference:
https://docs.aws.amazon.com/sagemaker/latest/APIReference/API_CreateDomain.html
https://docs.aws.amazon.com/sagemaker/latest/APIReference/API_UpdateDomain.html

EnableDockerAccess
Indicates whether the domain can access Docker.
Type: String
Valid Values: ENABLED | DISABLED

VpcOnlyTrustedAccounts
The list of AWS accounts that are trusted when the domain is created in VPC-only mode.
Type: Array of strings
Length Constraints: Fixed length of 12.
Pattern: ^\d+$

### Would you like to implement a fix?

No

### LINKED ISSUE #35382: [Bug]: sagemaker_domain r_studio_server_pro_app_settings reports drift on every apply if disabled
### Terraform Core Version

1.4.6, 1.7.0

### AWS Provider Version

5.1.0, 5.26.0

### Affected Resource(s)

r_studio_server_pro_app_settings for aws_sagemaker_domain/default_user_settings, aws_sagemaker_user_profile/user_settings

### Expected Behavior

When setting the r_studio_server_pro_app_settings to disabled for the resource aws_sagemaker_user_profile or the resource aws_sagemaker_domain, Terraform should not try to set the r_studio_server_pro_app_settings user_group. Running multiple applies with no code changes should not detect drift regardless.

### Actual Behavior

When setting the r_studio_server_pro_app_settings to disabled for the resource aws_sagemaker_user_profile or the resource aws_sagemaker_domain, Terraform repeatedly tries to set the user_group, even without code changes. If you don't set the r_studio_server_pro_app_settings at all you get a similar behavior where TF tries to "reset" access_status = "DISABLED" to null every time Terraform is run.Running back to back plan/apply with r_studio_server_pro_app_settings access_status set to "disabled"  results in a "change" detected for each user and the domain:

aws_sagemaker_domain.sagemaker_domain output:

```
default_user_settings {

          ~ r_studio_server_pro_app_settings {
              + user_group    = "R_STUDIO_USER"
                # (1 unchanged attribute hidden)
            }
```

sagemaker_domain.aws_sagemaker_user_profile.default_user output:
  
```
user_settings {
          
          ~ r_studio_server_pro_app_settings {
              + user_group    = "R_STUDIO_USER"
                # (1 unchanged attribute hidden)
            }
```

Not setting the r_studio_server_pro_app_settings at all results in multiple applies detect drift as well:
aws_sagemaker_domain.sagemaker_domain output:

```
 default_user_settings {
            # (2 unchanged attributes hidden)

          - r_studio_server_pro_app_settings {
              - access_status = "DISABLED" -> null
            }
```

sagemaker_domain.aws_sagemaker_user_profile.default_user output:
```
 user_settings {
          - r_studio_server_pro_app_settings {
              - access_status = "DISABLED" -> null
            }

```

### Relevant Error/Panic Output Snippet

_No response_

### Terraform Configuration Files

```
resource "aws_sagemaker_domain" "sagemaker_domain" {
  domain_name = var.domain_name
  auth_mode   = var.auth_mode
  vpc_id      = var.vpc_id
  subnet_ids  = var.subnet_ids

  default_user_settings {
    execution_role  = module.sagemaker_execution_roles.default_user_execution_role
    security_groups = [aws_security_group.sagemaker_sg.id]
    r_studio_server_pro_app_settings {
      access_status = "DISABLED"
    }

    jupyter_server_app_settings {
      default_resource_spec {
        sagemaker_image_arn = var.sagemaker_image_arn
        instance_type       = "system"
      }
    }
  }

  domain_settings {
    security_group_ids = [aws_security_group.sagemaker_sg.id]
  }

  kms_key_id              = module.efs_kms.kms_arn
  app_network_access_type = var.app_network_access_type

  retention_policy {
    home_efs_file_system = var.efs_retention_policy
  }
}

resource "aws_sagemaker_user_profile" "default_user" {
  for_each          = toset(var.user_profiles)
  domain_id         = aws_sagemaker_domain.sagemaker_domain.id
  user_profile_name = each.value

  user_settings {
    execution_role  = module.sagemaker_execution_roles.default_user_execution_role
    security_groups = [aws_security_group.sagemaker_sg.id]

    jupyter_server_app_settings {
      default_resource_spec {
        sagemaker_image_arn = var.sagemaker_image_arn
        instance_type       = "system"
      }
    }
    r_studio_server_pro_app_settings {
      access_status = "DISABLED"
    }
  }
}
```

### Steps to Reproduce

Run terraform plan, Terraform detects wanting to change the r_studio_server_pro_app_settings attribute user_group to "R_STUDIO_USER" even though the access_status is set to "DISABLED". Letting Terraform apply the "changes" then running apply again after no code changes ends in same result, Terraform detects a change to the r_studio_server_pro_app_settings attribute user_group to "R_STUDIO_USER".

### Debug Output

_No response_

### Panic Output

_No response_

### Important Factoids

Updated to latest Terraform (1.7) and AWS provider (5.26) with same results. 

### References

https://github.com/hashicorp/terraform-provider-aws/issues/33034

### Would you like to implement a fix?

None

### PR DESCRIPTION
**Pull request by deepakbshetty**
### Description

Adds docker_settings to domain_settings under `aws_sagemaker_domain` resource.
Add/Fix r_studio_server_pro_domain_settings update under domain_settings update to avoid drift. 

### Relations

Closes #35115
Fixes #31912 

### References

[Create Domain API](https://docs.aws.amazon.com/sagemaker/latest/APIReference/API_CreateDomain.html)
[Update Domain API](https://docs.aws.amazon.com/sagemaker/latest/APIReference/API_UpdateDomain.html)

### Output from Acceptance Testing

```console
$ make testacc TESTS=TestAccSageMaker_serial/Domain/domainSettingsDockerSettingsUpdated PKG=sagemaker
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/sagemaker/... -v -count 1 -parallel 20 -run='TestAccSageMaker_serial/Domain/domainSettingsDockerSettingsUpdated'  -timeout 360m
=== RUN   TestAccSageMaker_serial
=== PAUSE TestAccSageMaker_serial
=== CONT  TestAccSageMaker_serial
=== RUN   TestAccSageMaker_serial/Domain
=== RUN   TestAccSageMaker_serial/Domain/domainSettingsDockerSettingsUpdated
--- PASS: TestAccSageMaker_serial (381.56s)
    --- PASS: TestAccSageMaker_serial/Domain (381.56s)
        --- PASS: TestAccSageMaker_serial/Domain/domainSettingsDockerSettingsUpdated (381.55s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/sagemaker  404.668s



 make testacc TESTS=TestAccSageMaker_serial/Domain/rStudioServerProAppSettings PKG=sagemaker
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/sagemaker/... -v -count 1 -parallel 20 -run='TestAccSageMaker_serial/Domain/rStudioServerProAppSettings'  -timeout 360m
=== RUN   TestAccSageMaker_serial
=== PAUSE TestAccSageMaker_serial
=== CONT  TestAccSageMaker_serial
=== RUN   TestAccSageMaker_serial/Domain
=== RUN   TestAccSageMaker_serial/Domain/rStudioServerProAppSettings
--- PASS: TestAccSageMaker_serial (456.54s)
    --- PASS: TestAccSageMaker_serial/Domain (456.54s)
        --- PASS: TestAccSageMaker_serial/Domain/rStudioServerProAppSettings (456.54s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/sagemaker  479.221s



```

### CODE REVIEW

**Comment by deepakbshetty** [comment]
Hi @DrFaust92, appreciate if you can review the PR and provide feedback.

**Comment by DrFaust92** [comment]
[deepakbshetty](https://github.com/deepakbshetty) looks great. lets split to a PR per resource please

**Comment by deepakbshetty** [comment]
Dear @DrFaust92 , Thanks for the speedy review. Split PRs to 3 - #35477 & #35479

**Comment by deepakbshetty** [comment]
Hi @DrFaust92, Gentle nudge to review again and initiate tests on behalf

**Comment by deepakbshetty** [comment]
Hi @DrFaust92, May I kindly request a review post-split.

Hi @ewbankkit, if we can consider it for the next release? The PR brings in fixes for Sagemaker domain update settings and adds in feature to enable docker_settings.

File: internal/service/sagemaker/domain.go [hunk 1]
```diff
@@ -680,7 +680,6 @@ func ResourceDomain() *schema.Resource {
 									"user_group": {
 										Type:         schema.TypeString,
 										Optional:     true,
-										Default:      sagemaker.RStudioServerProUserGroupRStudioUser,
```
Comment (DrFaust92) [inline internal/service/sagemaker/domain.go:683]: This is a potenially breaking change, @ewbankkit wdyt?

File: internal/service/sagemaker/domain.go [hunk 2]
```diff
@@ -1223,10 +1269,50 @@ func expandDomainSettingsUpdate(l []interface{}) *sagemaker.DomainSettingsForUpd
...
 
+	if v, ok := m["docker_settings"].([]interface{}); ok && len(v) > 0 {
+		config.DockerSettings = expandDockerSettings(v)
+	}
+
 	if v, ok := m["execution_role_identity_config"].(string); ok && v != "" {
 		config.ExecutionRoleIdentityConfig = aws.String(v)
 	}
 
+	if v, ok := m["security_group_ids"].(*schema.Set); ok && v.Len() > 0 {
```
Comment (DrFaust92) [inline internal/service/sagemaker/domain.go:1280]: add tests for updating these values please

**Review by DrFaust92 — CHANGES_REQUESTED** [review]
missing some tests

File: internal/service/sagemaker/domain.go (hunk 2 above)
Comment (deepakbshetty) [reply internal/service/sagemaker/domain.go:1280]: Thank you for speedy review @DrFaust92. Have streamlined existing test case and renamed from testAccDomain_securityGroup to testAccDomain_defaultUserSettingsSecurityGroupUpdated.
( Removed testAccDomainConfig_securityGroup1 function )


$ make testacc TESTS=TestAccSageMaker_serial/Domain/defaultUserSettingsSecurityGroupUpdated PKG=sagemaker
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/sagemaker/... -v -count 1 -parallel 20 -run='TestAccSageMaker_serial/Domain/defaultUserSettingsSecurityGroupUpdated'  -timeout 360m
=== RUN   TestAccSageMaker_serial
=== PAUSE TestAccSageMaker_serial
=== CONT  TestAccSageMaker_serial
=== RUN   TestAccSageMaker_serial/Domain
=== RUN   TestAccSageMaker_serial/Domain/defaultUserSettingsSecurityGroupUpdated
--- PASS: TestAccSageMaker_serial (398.48s)
    --- PASS: TestAccSageMaker_serial/Domain (398.48s)
        --- PASS: TestAccSageMaker_serial/Domain/defaultUserSettingsSecurityGroupUpdated (398.48s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/sagemaker  424.234s

File: internal/service/sagemaker/domain.go (hunk 1 above)
Comment (DrFaust92) [reply internal/service/sagemaker/domain.go:683]: @deepakbshetty lets remove this from this PR. i see there is a follow PR to remove this

Comment (deepakbshetty) [reply internal/service/sagemaker/domain.go]: @DrFaust92 Removed.

**Comment by adv4000** [comment]
Workaround to enable DockerSettings is to use `provisioner "local-exec"`
In `awscc_sagemaker_domain` docker settings also not working https://github.com/hashicorp/terraform-provider-awscc/issues/1851
```
resource "aws_sagemaker_domain" "this" {
  domain_name             = "sagemaker-domain"
  auth_mode               = "IAM"
  app_network_access_type = "VpcOnly"
  vpc_id                  = "vpc-xxxxxxxxx"
  subnet_ids              = ["subnet-111111","subnet-222222"]

  default_user_settings  {
    execution_role      = "arn:aws:iam::111111111:role/SageMakerROLE"
    security_groups     = ["sg-xxxxxxxx"]
    default_landing_uri = "studio::"
    studio_web_portal   = "ENABLED"
   }
}


resource "terraform_data" "enable_sm_dockersettings" {
  provisioner "local-exec" {
    command = "aws sagemaker update-domain --domain-id ${aws_sagemaker_domain.this.id} --domain-settings-for-update {\"DockerSettings\":{\"EnableDockerAccess\":\"ENABLED\"}}"
  }
  depends_on = [aws_sagemaker_domain.this]
}

```

**Comment by adv4000** [comment]
Under Windows:
```
  provisioner "local-exec" {
    command = "aws sagemaker update-domain --domain-id ${aws_sagemaker_domain.this.id} --domain-settings-for-update {\"DockerSettings\":{\"EnableDockerAccess\":\"ENABLED\"}}"
  }
 ```
 Under Linux:
 ```
   provisioner "local-exec" {
    command = "aws sagemaker update-domain --domain-id ${aws_sagemaker_domain.this.id} --domain-settings-for-update `{\"DockerSettings\":{\"EnableDockerAccess\":\"ENABLED\"}}`"
  }
  ```

**Review by DrFaust92 — APPROVED** [review]
LGTM

**Comment by DrFaust92** [comment]
[deepakbshetty](https://github.com/deepakbshetty) can you rebase/resolve conflicts?

**Comment by rileyhun** [comment]
Any ETA on when this will be merged?

**Comment by deepakbshetty** [comment]
Hi @DrFaust92 , Rebased and merge conflicts solved

File: internal/service/sagemaker/sagemaker_test.go [hunk 3]
```diff
@@ -46,6 +46,29 @@ func TestAccSageMaker_serial(t *testing.T) {
 			"kernelGatewayAppSettings":                 testAccDomain_kernelGatewayAppSettings,
 			"kernelGatewayAppSettings_customImage":     testAccDomain_kernelGatewayAppSettings_customImage,
 			"kernelGatewayAppSettings_lifecycleConfig": testAccDomain_kernelGatewayAppSettings_lifecycleConfig,
+			"kernelGatewayAppSettings_defaultResourceAndCustomImage": testAccDomain_kernelGatewayAppSettings_defaultResourceSpecAndCustomImage,
```
Comment (DrFaust92) [inline internal/service/sagemaker/sagemaker_test.go:49]: something looks incorrect here with this list addition, it should be one addition and one rename

**Review by DrFaust92 — CHANGES_REQUESTED** [review]

File: internal/service/sagemaker/sagemaker_test.go (hunk 3 above)
Comment (deepakbshetty) [reply internal/service/sagemaker/sagemaker_test.go:49]: Sorry, fixed now.

**Review by DrFaust92 — APPROVED** [review]
LGTM (didnt validate tests)

**Review by DrFaust92 — APPROVED** [review]

**Comment by deepakbshetty** [comment]
Hi @jar-b May I request if you can trigger tests again

**Comment by deepakbshetty** [comment]
Hi @ewbankkit, If I may request this for merge. There is considerable upvote to include this feature

**Review by DrFaust92 — APPROVED** [review]
LGTM

**Comment by deepakbshetty** [comment]
Dear @ewbankkit, Any chance if you consider these sagemaker domain requests that have passed checks #35416 and #38457 and clears multiple issues

**Comment by deepakbshetty** [comment]
> Dear @ewbankkit, Any chance if you consider these sagemaker domain requests that have passed checks #35416 and #38457 and clears multiple issues

@ewbankkit For your kind consideration of both PRs for sagemaker domain.

**Comment by jmeisele** [comment]
Hello @deepakbshetty , any ideas on resolving these conflicts so we can get this PR merged? Same with [web portal settings PR](https://github.com/hashicorp/terraform-provider-aws/pull/38457)

**Comment by deepakbshetty** [comment]
> Hello @deepakbshetty , any ideas on resolving these conflicts so we can get this PR merged? Same with [web portal settings PR](https://github.com/hashicorp/terraform-provider-aws/pull/38457)

Hi 👋  @jmeisele, These seem to have appeared due to PR aws sdk go v2 migration for sagemaker. I have a look this in the weekend. Would have been simpler if this code was merged prior.

**Comment by deepakbshetty** [comment]
> > Hello @deepakbshetty , any ideas on resolving these conflicts so we can get this PR merged? Same with [web portal settings PR](https://github.com/hashicorp/terraform-provider-aws/pull/38457)
> 
> Hi 👋 @jmeisele, These seem to have appeared due to PR aws sdk go v2 migration for sagemaker. I have a look this in the weekend. Would have been simpler if this code was merged prior.

Hi @jmeisele, Resolved merge conflicts and both PRs are now inline with aws-go-sdk-v2 changes.

**Comment by deepakbshetty** [comment]
Hi @ewbankkit, If you can kindly consider PRs #35416 an #38457 for sagemaker_domain. Not having these settings block is causing us to fall back non terraform approach

**Comment by deepakbshetty** [comment]
Hi @ewbankkit Is there any opportunity consider this long standing PR for current provider release please ?

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

```console
% make testacc TESTARGS='-run=TestAccSageMaker_serial/Domain/basic\|TestAccSageMaker_serial/Domain/domainSettings' PKG=sagemaker
make: Verifying source code with gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go1.22.6 test ./internal/service/sagemaker/... -v -count 1 -parallel 20  -run=TestAccSageMaker_serial/Domain/basic\|TestAccSageMaker_serial/Domain/domainSettings -timeout 360m
=== RUN   TestAccSageMaker_serial
=== PAUSE TestAccSageMaker_serial
=== CONT  TestAccSageMaker_serial
=== RUN   TestAccSageMaker_serial/Domain
=== RUN   TestAccSageMaker_serial/Domain/basic
=== RUN   TestAccSageMaker_serial/Domain/domainSettings
=== RUN   TestAccSageMaker_serial/Domain/domainSettingsDockerSettingsUpdated
--- PASS: TestAccSageMaker_serial (745.01s)
    --- PASS: TestAccSageMaker_serial/Domain (745.01s)
        --- PASS: TestAccSageMaker_serial/Domain/basic (250.36s)
        --- PASS: TestAccSageMaker_serial/Domain/domainSettings (262.40s)
        --- PASS: TestAccSageMaker_serial/Domain/domainSettingsDockerSettingsUpdated (232.24s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/sagemaker	750.202s
```

**Comment by ewbankkit** [comment]
@deepakbshetty Thanks for the contribution :tada: :clap:.

**Pull request merged**